File: cpp/sort/sort.cpp
#include "sort.h"

int main(void) {
	test(new BubbleSort());
	test(new SelectSort());
	test(new InsertSort());
	test(new ShellSort());
	test(new MergeSort());
	test(new QuickSort());
	test(new HeapSort());
	test(new RadixSort());
	exit(EXIT_SUCCESS);
	return 0;
}

void test(Sort* s) {
	IntArray* arr = new IntArray(20, 1000);
	s->sort(*arr);
	printf("%s\t", s->getClassName());
	arr->print();
	delete arr;
}

// implement IntArray
IntArray::IntArray(int size) : length(size) {
	arr = (int*) malloc(sizeof(int) * size);
	for(int i = 0; i < size; i++) arr[i] = 0;
}
IntArray::IntArray(int size, int digit) : length(size) {
	arr = (int*) malloc(sizeof(int) * size);
	srand(time(0)-RANDOM(digit));
	for(int i = 0; i < size; i++) {
		arr[i] = RANDOM(digit);
	}
}
IntArray::~IntArray() {
	free(arr);
}
int& IntArray::operator[](int i) {
    return arr[i];
}
int IntArray::size() {
	return length;
}
void IntArray::print() {
	for(int i = 0; i < length; i++) {
		printf((i == 0 ? "[" : ""));
		printf("%d", arr[i]);
		printf((i == length-1 ? "]\n" : ", "));
	}
}

const char* Sort::getClassName(){
	const char* name = typeid(*this).name();
	while(1) {
		if((*name) > '9') return name;
		name++;
	}
}

// 冒泡排序
void BubbleSort::sort(IntArray arr) {
	for (int i = arr.size() - 1; i >= 0; i--) {
		for (int j = 0; j < i; j++) {
			if (arr[j] > arr[j + 1]) {
				int temp = arr[j + 1];
				arr[j + 1] = arr[j];
				arr[j] = temp;
			}
		}
	}
}
// 选择排序
void SelectSort::sort(IntArray arr) {
	for (int i = 0, n = arr.size(); i < n; i++) {
		int min = i;
		for (int j = i + 1; j < n; j++) {
			if (arr[j] < arr[min])
				min = j;
		}
		if (min != i) {
			int temp = arr[i];
			arr[i] = arr[min];
			arr[min] = temp;
		}
	}
}
// 插入排序
void InsertSort::sort(IntArray arr) {
	for (int i = 1, n = arr.size(); i < n; i++) {
		int j = i - 1;
		int tmp = arr[i];
		while (j >= 0 && tmp < arr[j]) {
			arr[j + 1] = arr[j];
			j--;
		}
		arr[j + 1] = tmp;
	}
}
// 希尔排序
void ShellSort::sort(IntArray arr) {
	int n = arr.size();
	int gap = n / 2;
	while (1 <= gap) {
		for (int i = gap; i < n; i++) {
			int j = i - gap;
			int tmp = arr[i];
			while (j >= 0 && tmp < arr[j]) {
				arr[j + gap] = arr[j];
				j -= gap;
			}
			arr[j + gap] = tmp;
		}
		gap /= 2;
	}
}
// 归并排序
void MergeSort::sort(IntArray arr) {
	sort(arr, 0, arr.size() - 1);
}
void MergeSort::sort(IntArray arr, int l, int r) {
	if (l >= r) return;
	int m = (l + r) / 2;
	sort(arr, l, m);
	sort(arr, m + 1, r);
	merge(arr, l, m, r);
}
void MergeSort::merge(IntArray arr, int l, int m, int r) {
	IntArray tmp(r - l + 1);
	int i = l, j = m + 1, k = 0;
	while (i <= m && j <= r) tmp[k++] = (arr[i] < arr[j]) ? arr[i++] : arr[j++];
	while (i <= m) tmp[k++] = arr[i++];
	while (j <= r) tmp[k++] = arr[j++];
	for(int i = l; i <= r; i++) arr[i] = tmp[i-l];
}
// 快速排序
void QuickSort::sort(IntArray arr) {
	sort(arr, 0, arr.size() - 1);
}
void QuickSort::sort(IntArray arr, int l, int r) {
	if (l >= r) return;
	int m = partition(arr, l, r);
	sort(arr, l, m - 1);
	sort(arr, m + 1, r);
}
int QuickSort::partition(IntArray arr, int l, int r) {
	int m = l;
	for (int u = l; u < r; u++) {
		if (arr[u] <= arr[r]) {
			int temp = arr[u];
			arr[u] = arr[m];
			arr[m] = temp;
			m++;
		}
	}
	int temp = arr[r];
	arr[r] = arr[m];
	arr[m] = temp;
	return m;
}
// 堆排序
void HeapSort::sort(IntArray arr) {
	int n = arr.size() - 1;
	for (int i = (n / 2); i >= 0; i--) sift(arr, i, n);
	for (int i = n; i >= 1; i--) {
		int temp = arr[i];
		arr[i] = arr[0];
		arr[0] = temp;
		sift(arr, 0, i - 1);
	}
}
void HeapSort::sift(IntArray arr, int l, int r) {
	int i = l, j = l * 2 + 1, tmp = arr[l];
	while (j <= r) {
		if (j < r && arr[j] < arr[j + 1]) j++;
		if (tmp < arr[j]) {
			arr[i] = arr[j];
			i = j;
			j = 2 * i + 1;
		} else break;
	}
	arr[i] = tmp;
}
// 基数排序
void RadixSort::sort(IntArray arr) {
	int d = maxbit(arr), radix = 1, n = arr.size();
	IntArray tmp(arr.size());
	for (int i = 1; i <= d; i++) {
		IntArray count(10);
		for (int j = 0; j < n; j++) count[(arr[j] / radix) % 10]++;
		for (int j = 1; j < 10; j++) count[j] += count[j - 1];
		for (int j = n - 1; j >= 0; j--) tmp[--count[(arr[j] / radix) % 10]] = arr[j];
		for (int j = 0; j < n; j++) arr[j] = tmp[j];
		radix *= 10;
	}
}
int RadixSort::maxbit(IntArray arr) {
	int d = 1, p = 10, n = arr.size();
	for (int i = 0; i < n; i++) {
		int v = arr[i];
		while (v >= p) {
			p *= 10;
			++d;
		}
	}
	return d;
}

File: go/tool/sort.go
package main

import (
	"bytes"
	"fmt"
	"reflect"
	"strings"
	"time"
	"math/rand"
	"strconv"
)

var rdm *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))

func array(size int, digit int) []int {
	arr := make([]int, size, size)
	for i := range arr {
		arr[i] = rdm.Intn(digit)
	}
	return arr
}
func getSimpleName(i interface{}) string {
	name := reflect.ValueOf(i).Type().String()
	a := strings.Split(name, ".")
	return a[len(a) - 1]
}
func getArrayString(arr []int) string {
	buf := bytes.Buffer{}
	buf.WriteString("[")
	for i := 0; i < len(arr) - 1; i++ {
		buf.WriteString(strconv.Itoa(arr[i]))
		buf.WriteString(", ")
	}
	buf.WriteString(strconv.Itoa(arr[len(arr) - 1]))
	buf.WriteString("]")
	return buf.String()
}
func test(sort Sorter) {
	arr := array(20, 1000)
	sort.sort(arr)
	fmt.Printf("%s\t%s\n", getSimpleName(sort), getArrayString(arr))
}

type Sorter interface {
	sort(arr []int)
}

type BubbleSort struct{}

func (self BubbleSort) sort(arr []int) {
	for i := len(arr) - 1; i >= 0; i-- {
		for j := 0; j < i; j++ {
			if arr[j] > arr[j + 1] {
				tmp := arr[j]
				arr[j] = arr[j + 1]
				arr[j + 1] = tmp
			}
		}
	}
}

type SelectSort struct{}

func (self SelectSort) sort(arr []int) {
	count := len(arr)
	for i := 0; i < count; i++ {
		min := i
		for j := i + 1; j < count; j++ {
			if arr[min] > arr[j] {
				min = j
			}
		}
		if min != i {
			tmp := arr[i]
			arr[i] = arr[min]
			arr[min] = tmp
		}
	}
}

type InsertSort struct{}

func (self InsertSort) sort(arr []int) {
	count := len(arr)
	for i := 1; i < count; i++ {
		j := i - 1
		tmp := arr[i]
		for j >= 0 && arr[j] > tmp {
			arr[j+1] = arr[j]
			j -= 1
		}
		arr[j+1] = tmp
	}
}

type ShellSort struct{}

func (self ShellSort) sort(arr []int) {
	count := len(arr)
	gap := count / 2
	for 1 <= gap {
		for i := gap; i < count; i++ {
			j := i - gap
			tmp := arr[i]
			for j >= 0 && arr[j] > tmp {
				arr[j+gap] = arr[j]
				j -= gap
			}
			arr[j+gap] = tmp
		}
		gap /= 2
	}
}

type MergeSort struct{}

func (self MergeSort) sort(arr []int) {
	self.recursive(arr, 0, len(arr)-1)
}

func (self MergeSort) recursive(arr []int, l int, r int)  {
	if l >= r { return }
	m := (r + l) / 2
	self.recursive(arr, l, m)
	self.recursive(arr, m+1, r)
	self.merge(arr, l, m, r)
}

func (self MergeSort) merge(arr []int, l int, m int, r int)  {
	tmp := make([]int, 0, r - l + 1)
	i, j := l, m+1
	for i <= m && j <= r {
		if arr[i] < arr[j] {
			tmp = append(tmp, arr[i])
			i++
		} else {
			tmp = append(tmp, arr[j])
			j++
		}
	}
	for i <= m {
		tmp = append(tmp, arr[i])
		i++
	}
	for j <= r {
		tmp = append(tmp, arr[j])
		j++
	}
	for i, v := range tmp {
		arr[l+i] = v;
	}
}

type QuickSort struct{}

func (self QuickSort) sort(arr []int) {
	self.recursive(arr, 0, len(arr)-1)
}

func (self QuickSort) recursive(arr []int, l int, r int)  {
	if l >= r { return }
	m := self.partition(arr, l, r)
	self.recursive(arr, l, m-1)
	self.recursive(arr, m+1, r)
}

func (self QuickSort) partition(arr []int, l int, r int) int {
	m := l
	for i := l; i < r; i++ {
		if arr[i] < arr[r] {
			temp := arr[i]
			arr[i] = arr[m]
			arr[m] = temp
			m++
		}
	}
	temp := arr[r]
	arr[r] = arr[m]
	arr[m] = temp
	return m
}

type HeapSort struct{}

func (self HeapSort) sort(arr []int) {
	n := len(arr) - 1
	for i := (n / 2); i >= 0; i-- {
		self.sift(arr, i, n)
	}
	for i := n; i >= 1; i-- {
		temp := arr[i]
		arr[i] = arr[0]
		arr[0] = temp
		self.sift(arr, 0, i - 1)
	}
}

func (self HeapSort) sift(arr []int, l int, r int) {
	i, j, tmp := l, l * 2 + 1, arr[l]
	for j <= r {
		if j < r && arr[j] < arr[j + 1] {j++}
		if tmp < arr[j] {
			arr[i] = arr[j]
			i = j
			j = 2 * i + 1
		} else {break}
	}
	arr[i] = tmp
}

type RadixSort struct{}

func (self RadixSort) sort(arr []int) {
	n := len(arr)
	d, radix, tmp := self.maxbit(arr), 1, make([]int, n, n)
	for i := 1; i <= d; i++ {
		count := make([]int, 10, 10)
		for j := 0; j < n; j++ {
			count[(arr[j] / radix) % 10]++
		}
		for j := 1; j < 10; j++ {
			count[j] += count[j - 1]
		}
		for j := n-1; j >= 0; j-- {
			k := (arr[j] / radix) % 10
			count[k] -= 1
			tmp[count[k]] = arr[j]
		}
		for j := 0; j < n; j++ {
			arr[j] = tmp[j]
		}
		radix *= 10
	}
}

func (self RadixSort) maxbit(arr []int) int {
	d, p := 1, 10
	for _, i := range arr {
		for i >= p {
			p *= 10
			d++
		}
	}
	return d;
}

func main() {
	test(BubbleSort{})
	test(SelectSort{})
	test(InsertSort{})
	test(ShellSort{})
	test(MergeSort{})
	test(QuickSort{})
	test(HeapSort{})
	test(RadixSort{})
}
